### PR TITLE
Support `#[track_caller]` on closures.

### LINKED ISSUE #74042: #[track_caller] on closures?
One problem I have is that it's not possible to put `#[track_caller]` on closures, this is really useful when `thread_local` is involved because then there's a lot of logic in closures that you might want to propagate
ie

```rust
thread_local! {static DONE: bool = false;}

#[track_caller]
fn assert_done() {
    DONE.with(
        #[track_caller]
        |b| assert!(b),
    );
}

fn main() {
    assert_done();
}
```

_Originally posted by @elichai in https://github.com/rust-lang/rust/issues/47809#issuecomment-653777647_

### PR DESCRIPTION
**Pull request by anp**
Closes #74042.

### CODE REVIEW

**Comment by anp** [comment]
Fails locally with:

```
thread 'rustc' panicked at 'assertion failed: `(left == right)`
  left: `2`,
 right: `3`: #[track_caller] fn's must have 1 more argument in their ABI than in their MIR'
```

**Comment by lcnr** [comment]
afaik closures use `FnOnce::call_once`, which has the "rust-call" calling convention.

This calling convention takes the argument list and wraps them in a tuple to support generics.
Not sure what has to be done about that here though :thinking:

**Comment by anp** [comment]
Thanks! That jogged my memory for the portion of codegen to look at. Since the assertion is there to prevent miscompilation I tried commenting it out and the argument passing appeared to "just work". Will still need to figure out the right number to assert on in the presence of untupled args but that can be done later.

The next issue is that the "definition span" we get for a closure-as-fn-ptr appears to be in libcore:

```
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `"/home/adam/rust/src/libcore/ops/function.rs"`,
 right: `"/home/adam/rust/src/test/ui/rfc-2091-track-caller/track-caller-closure.rs"`', /home/adam/rust/src/test/ui/rfc-2091-track-caller/track-caller-closure.rs:52:5
```

I think this will need to be modified in https://github.com/rust-lang/rust/blob/2210abea71270867fe2c69782f282c654e106fac/src/librustc_mir/interpret/intrinsics/caller_location.rs#L17 but I might be wrong and it might also require special closure logic when resolving fn ptrs in https://github.com/rust-lang/rust/blob/2666aed4989c3bec9cf9f94b2d15beda4e5407f7/src/librustc_middle/ty/instance.rs#L375.

cc @eddyb btw

**Comment by eddyb** [comment]
Sounds like you need to make some of the `Fn`-trait-related shims pretend to have `#[track_caller]` depending on whether the underlying function/closure does. I didn't think about that when I suggested closures "should just work" but:

You might be able to get away with testing a case where a closure doesn't implement more than `FnOnce` (e.g. capture something like a `String` that is entirely consumed by the closure body), because that shouldn't get any shims.

Then again, you mention coercing to a `fn` pointer being the only case that doesn't work, I'm not sure what the best solution for that is, maybe checking that we're in a reify shim and doing `Instance::resolve` on the `DefId`/`Substs` pair of the reify shim? That way, we may get back an `InstanceDef::Item` referring to the closure, or a shim that makes it clear what the closure `DefId` is?

File: src/librustc_codegen_ssa/mir/block.rs [hunk 1]
```diff
@@ -808,11 +808,12 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
...
-                fn_abi.args.len(),
-                args.len() + 1,
-                "#[track_caller] fn's must have 1 more argument in their ABI than in their MIR",
-            );
+            // FIXME figure out the right number of args for closures
+            // assert_eq!(
+            //     fn_abi.args.len(),
+            //     args.len() + 1,
+            //     "#[track_caller] fn's must have 1 more argument in their ABI than in their MIR",
+            // );
```
Comment (bjorn3) [inline src/librustc_codegen_ssa/mir/block.rs:816]: If the abi is "Rust-call" then the argument count is 1 + the amount of elements in the type of the second argument (which is a tuple).

**Comment by Dylan-DPC** [comment]
@anp any updates?

**Comment by Dylan-DPC** [comment]
@anp thanks for taking the time to contribute. Sadly this PR is inactive so I'm going to close it. If you wish and you have the time to complete this, you can open a new pull request and we'll take it from there.

**Pull request closed without merging**